### ISSUE TITLE
Weird behavior of tables.expression.Expr when slice is outside of data samples

### ISSUE THREAD

**Issue by jackdbd** [PyTables/PyTables#651]
I noticed a strange behavior when evaluating an expression on a slice which is outside of the data samples.

I would have expected to get an empty numpy array. Instead I got a single scalar. Is this a bug or the expected behavior?

Here is my environment:

```shell
Windows 10 64 bit
Python 2.7 32 bit
hdf5 1.8.18
tables 3.4.2
numexpr 2.6.4
numpy 1.13.1+mkl
```

I wrote these tests to show this issue.

```python
import tables
import unittest
import numpy as np


class TestNumexpr(unittest.TestCase):

    def test_slice(self):
        with tables.open_file("my_file.h5", mode="w", title="Test") as h5file:
            group = h5file.create_group(h5file.root, "columns", "Group Name")
            num_elements = 10
            arr_a = np.ones(num_elements)
            a = h5file.create_array(group, "a", arr_a, "Array a")
            arr_b = np.ones(num_elements) * 2.0
            b = h5file.create_array(group, "b", arr_b, "Array b")
            expr = tables.Expr("2 * a + b * 3")
            start, stop, step = (0, 7, None)
            expr.set_inputs_range(start, stop, step)
            result = expr.eval()
            expected = np.ones(stop) * 8
            self.assertListEqual(result.tolist(), expected.tolist())

    # This test fails: expr.eval() returns a scalar (8.0), not an empty np.array
    def test_empty_slice_for_binary_function_returns_empty_array(self):
        with tables.open_file("my_file.h5", mode="w", title="Test") as h5file:
            group = h5file.create_group(h5file.root, "columns", "Group Name")
            num_elements = 10
            arr_a = np.ones(num_elements)
            a = h5file.create_array(group, "a", arr_a, "Array a")
            arr_b = np.ones(num_elements) * 2.0
            b = h5file.create_array(group, "b", arr_b, "Array b")
            expr = tables.Expr("2 * a + b * 3")
            start, stop, step = (0, 0, None)
            # we get the same behavior if we pick a slice outside of the array
            # start, stop, step = (100, 1000, None)
            expr.set_inputs_range(start, stop, step)
            result = expr.eval()
            expected = np.array([])
            self.assertListEqual(result.tolist(), expected.tolist())

    # This test fails: expr.eval() returns a scalar (8.0), not an empty np.array
    def test_empty_slice_for_unary_function_returns_empty_array(self):
        with tables.open_file("my_file.h5", mode="w", title="Test") as h5file:
            group = h5file.create_group(h5file.root, "columns", "Group Name")
            num_elements = 10
            arr_a = np.ones(num_elements)
            a = h5file.create_array(group, "a", arr_a, "Array a")
            expr = tables.Expr("sin(a)")
            start, stop, step = (0, 0, None)
            # we get the same behavior if we pick a slice outside of the array
            # start, stop, step = (100, 1000, None)
            expr.set_inputs_range(start, stop, step)
            result = expr.eval()
            expected = np.array([])
            self.assertListEqual(result.tolist(), expected.tolist())


if __name__ == "__main__":
    unittest.main()
```

**Comment by FrancescAlted** [comment]
Yep, smells like a bug.  Interested in providing a fix?

**Comment by jackdbd** [comment]
I cloned the repo and installed everything on my machine.
I'm having a look at it right now. Line 579 in `expression.py` looks strange to me. Why return a value if there are no rows?

```python
if i_nrows == 0:
    # No elements to compute
    return self._single_row_out
```

I tried to replace it with the following, but 6 tests fail (I didn't run the test with the `--heavy` flag).

```python
if i_nrows == 0:
    # No elements to compute, but keep original array data type
    return np.array([], dtype=self._single_row_out.dtype)
```

Here is my machine (it's a different computer, it's not the same as the one I mentioned at the beginning).

```shell
Xubuntu 16.04.3 LTS (64 bit)
Python 3.5.2 (virtual environment)
```

**Comment by FrancescAlted** [comment]
Looks like this was fixed by #674.  Closing.
